src/Workspaces/Remote/Core/RemoteHostAssetSerialization.cs

```diff
@@ -3,69 +3,107 @@
-            // receiver can assert that both sides are talking about the same solution snapshot and no weird invariant
-            // breaks have occurred.
-            solutionChecksum.WriteTo(writer);
+            var foundChecksumCount = 0;
 
-            // special case
-            if (checksums.Length == 0)
-                return;
+            await scope.AddAssetsAsync(
+                assetPath,
+                checksums,
+                WriteAssetToPipeAsync,
+                cancellationToken).ConfigureAwait(false);
 
-            Debug.Assert(assetMap != null);
+            Contract.ThrowIfTrue(foundChecksumCount != checksums.Length);
 
-            for (var i = 0; i < checksums.Span.Length; i++)
+            return;
+
+            async ValueTask WriteAssetToPipeAsync(Checksum checksum, object asset, CancellationToken cancellationToken)
             {
-                var checksum = checksums.Span[i];
-                var asset = assetMap[checksum];
-
-                // We flush after each item as that forms a reasonably sized chunk of data to want to then send over the
-                // pipe for the reader on the other side to read.  This allows the item-writing to remain entirely
-                // synchronous without any blocking on async flushing, while also ensuring that we're not buffering the
-                // entire stream of data into the pipe before it gets sent to the other side.
-                WriteAsset(writer, serializer, context, asset, cancellationToken);
-                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+                Contract.ThrowIfNull(asset);
+                foundChecksumCount++;
+
+                using var pooledObject = s_streamPool.GetPooledObject();
+                var tempStream = pooledObject.Object;
+                tempStream.Position = 0;
+
+                // Don't truncate the stream as we're going to be writing to it multiple times.  This will allow us to
+                // reuse the internal chunks of the buffer, without having to reallocate them over and over again.
+                tempStream.SetLength(0, truncate: false);
+
+                // Write the asset to a temporary buffer so we can calculate its length.  Note: as this is an in-memory
+                // temporary buffer, we don't have to worry about synchronous writes on it blocking on the pipe-writer.
+                // Instead, we'll handle the pipe-writing ourselves afterwards in a completely async fasion.
+                WriteAssetToTempStream(tempStream, checksum, asset);
+
+                // Write the length of the asset to the pipe writer so the reader knows how much data to read.
+                WriteLengthToPipeWriter(tempStream.Length);
+
+                // Ensure we flush out the length so the reading side knows how much data to read.
+                await pipeWriterStream.FlushAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
the nice thing about this approach is that we can be all async here where we need to be (when actually doing the IO).  But the ObjecReader/ObjectWriter parts can all be synchronous on top of streams (without doing any blocking on any actual cross process IO/async).